src/fm.py

```diff
@@ -1,37 +1,208 @@
+            argparser.print_help()
+            sys.exit(1)
+
+def fasta_translator(input_file):
+    output_dict = {}
+    start = True
+    for i in input_file:
+        i = i.strip()
+        if len(i) == 0:
+            continue
+        if i[0] == ">":
+            if  not start:
+                output_dict[name] = seq + "$"
+            name = i[1:].strip()
+            seq = ""
+            if start:
+                start = False
+        elif start:
+            continue
+        else:
+            seq += i
+    output_dict[name] = seq + "$"
+    return output_dict
+
+def fastq_translator(input_file):
+    output_dict = {}
+    for i in input_file:
+        if i[0] == "@":
+            name = i[1:].strip()
+        else:
+            seq = i.strip()
+            output_dict[name] = seq
+    return(output_dict)
+
+def compute_buckets(SA, reference):
+    bucket_len = {}
+    for i in SA:
+        if reference[i - 1] in bucket_len:
+            bucket_len[reference[i - 1]] += 1
+        else:
+            bucket_len[reference[i - 1]] = 1
+    bucket_len = dict(sorted(bucket_len.items()))
+    buckets = {}
+    len_traversed = 0
+
+    for key in bucket_len:
+        temp = len_traversed
+        len_traversed += bucket_len[key]
+        buckets[key] = temp
+
+    return buckets
+
+def create_tracker(buckets, SA):
+    tracker = {}
+    for i in buckets:
+        tracker[i] = [None for i in range(len(SA) + 1)]
+    return tracker
+
+def create_ranks(buckets, SA, reference):
+    tracker = create_tracker(buckets, SA)
+    for char in tracker:
+        tracker[char][0] = 0
+    for i in range(1, len(SA) + 1):
+        for char in tracker:
+            tracker[char][i] = tracker[char][i - 1]
+        prev_bwt_char = reference[SA[i - 1] - 1]
+        tracker[prev_bwt_char][i] += 1
+    return tracker
+
+def SA_naive(string):
+    SA = [i for i in range(len(string))]
+    SA = [string[SA[int(i)]:] for i in SA]
+    SA.sort()
+    SA = [len(string) - len(i) for i in SA]
+    return(SA)
+
+def create_BWT_structures(ref_file, name):
+    fasta_dict = fasta_translator(ref_file)
+
+    SA_file = open(name + r"_SA.txt", "w")
+    bucket_file = open(name + r"_buckets.txt", "w")
+    rank_file = open(name + r"_rank.txt", "w")
+
+    for key in fasta_dict:
+        SA = SA_naive(fasta_dict[key])
+        SA_print = ",".join(str(i) for i in SA)
+        SA_file.write(SA_print)
+        buckets = compute_buckets(SA, fasta_dict[key])
+        for bucket in buckets:
+            bucket_file.write(bucket + " " + str(buckets[bucket]) + "\t")
+        rank = create_ranks(buckets, SA, fasta_dict[key])
+        rank_chars = list(rank.keys())
+        rank_file.write(",".join(rank_chars))
+        for i in range(len(SA) + 1):
+            rank_file.write("\t")
+            row = [str(rank[char][i]) for char in rank_chars]
+            rank_file.write(",".join(row))
+        rank_file.write("\n")
+        SA_file.write("\n")
+        bucket_file.write("\n")

```

Comment:
Der er for så vidt ikke noget galt i at skrive sin egen serialisering, men Python har allerede et bibliotek, `pickle`, der kan gøre det for dig, mere generelt, og en del hurtigere.